File: src/solutions/03/ts/03.ts
import { join } from "path";
import { cloneDeep } from "lodash";
import { readInputFile } from "../../../lib/input";

interface Coord {
  x: number;
  y: number;
}

interface State {
  grid: { [key: string]: boolean };
  treesCount: number;
  currentPos: Coord;
  currentGridWidth: number;
}

const coordToKey = ({ x, y }: Coord) => `${x},${y}`;
const keyToCoord = (k: string): Coord => {
  const [x, y] = k.split(",").map(Number);
  return { x, y };
};

let gridWidth = 0;
let gridHeight = 0;
const sourceGrid: { [key: string]: boolean } = {};

function calculateNext(slope: [number, number], state: State): State {
  let { treesCount, grid, currentPos, currentGridWidth } = state;
  const [TARGET_X, TARGET_Y] = slope;

  const nextPos: Coord = { x: currentPos.x + TARGET_X, y: currentPos.y + TARGET_Y };
  // if (nextPos.y + 1 > gridHeight) return;
  if (nextPos.x + 1 > currentGridWidth) {
    // copy grid right
    Object.entries(sourceGrid).forEach(([pos, val]) => {
      const coordPos = keyToCoord(pos);
      coordPos.x = coordPos.x + currentGridWidth;
      grid[coordToKey(coordPos)] = val;
    });
    currentGridWidth += gridWidth;
  }
  if (grid[coordToKey(nextPos)]) {
    treesCount += 1;
  }
  currentPos.x = nextPos.x;
  currentPos.y = nextPos.y;
  return { currentGridWidth, currentPos, grid, treesCount };
}

function createSourceGrid(input: string) {
  input.split("\n").forEach((line, y) => {
    if (y + 1 > gridHeight) {
      gridHeight = y + 1;
    }
    line.split("").forEach((spot, x) => {
      if (x + 1 > gridWidth) {
        gridWidth = x + 1;
      }
      sourceGrid[coordToKey({ x, y })] = spot === "#";
    });
  });
}

function countTreesInSlope(slope: [number, number]) {
  let grid = cloneDeep(sourceGrid);
  let currentGridWidth = gridWidth;
  let treesCount = 0;
  let currentPos: Coord = { x: 0, y: 0 };

  let state: State = { grid, currentGridWidth, treesCount, currentPos };

  while (state.currentPos.y + slope[1] + 1 <= gridHeight) {
    state = calculateNext(slope, state);
  }
  return state.treesCount;
}

const SLOPES: [number, number][] = [
  [1, 1],
  [3, 1],
  [5, 1],
  [7, 1],
  [1, 2],
];

async function run() {
  const input = await readInputFile(join(__dirname, "../data.txt"));
  createSourceGrid(input);

  const part1 = countTreesInSlope([3, 1]);
  console.log(`Part 1: ${part1}`);

  const resultsForSlopes = SLOPES.map((slope) => countTreesInSlope(slope));
  const part2 = resultsForSlopes.reduce((total, num) => total * num, 1);
  console.log(`Part 2: ${part2}`);
}

run();

File: src/solutions/03/rust/03.rs
use crate::lib::input::load_input;
use std::collections::HashMap;

type Coord = (u32, u32);
type Grid = HashMap<Coord, bool>;
type GridInfo = (Grid, u32, u32);

fn create_source_grid(input: String) -> GridInfo {
    let mut source_height: u32 = 0;
    let mut source_width: u32 = 0;
    let mut grid: Grid = Grid::new();

    for (y, line) in input.lines().enumerate() {
        if (y as u32) + 1 > source_height {
            source_height = (y as u32) + 1;
        }
        for (x, spot) in line.split("").enumerate() {
            if (x as u32) + 1 > source_width {
                source_width = (x as u32) + 1;
            }
            let spot_value = if spot == "#" { true } else { false };
            grid.insert((x as u32, y as u32), spot_value);
        }
    }

    return (grid, source_width, source_height);
}

fn count_slope(slope: (u32, u32), source_grid_info: GridInfo) {
    let trees_count: u32 = 0;
    let current_pos: Coord = (0, 0);
}

pub fn run() {
    let input = load_input("src/solutions/03/data.txt");
    let grid_info = create_source_grid(input);
    let slopes = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)];
}
